keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1131 @@
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Reduce CPU time and heap churn of encode and decode for cases that can't use
+  Protobuf:
+  - custom resource storage
+  - custom resource serving
+  - dynamic clients
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Replace existing usage of Protobuf.
+- Substantially reduce the size of encoded objects (a modest size reduction is
+  anticipated).
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Format
+
+The output of the CBOR encoder is a single tagged data item as specified in
+“[Self-Described
+CBOR](https://www.rfc-editor.org/rfc/rfc8949.html#name-self-described-cbor)”,
+with no additional envelope. Self-described CBOR -- a tagged data item with tag
+number 55799 -- has the same semantics as the same data item with no tag, with
+the convenient property that its encoded form is always prefixed by
+`0xd9d9f7`. By design, this prefix is never found at the beginning of a JSON
+text and can be used as a “magic number” to distinguish the data format of a
+stored object at rest.
+
+To support decoding custom resources that have been stored as a mixture of JSON
+and CBOR, the CBOR serializer will implement RecognizingDecoder by checking for
+the prefix `0xd9d9f7`.
+
+Streaming responses (i.e. watches) will be serialized as CBOR Sequences. A [CBOR
+Sequence](https://www.rfc-editor.org/rfc/rfc8742.html) is a concatenation of
+zero or more CBOR data items, with no additional framing. This is effectively
+equivalent to the existing JSON stream serialization behavior and takes
+advantage of the property that, like JSON objects – and unlike non-object JSON
+documents, e.g. numbers – CBOR data items are self-delimiting.
+
+### Negotiation
+
+Proactive content negotiation will be supported for clients that want to receive
+CBOR-encoded responses using the MIME type “application/cbor” in the Accept
+request header. For compatibility with API servers that don’t support CBOR,
+clients should also accept “application/json” (with a lower quality factor) and
+choose the appropriate decoder based on the Content-Type response header.
+
+Similarly, streaming requests should use the MIME type for CBOR Sequences,
+“application/cbor-seq”.
```

Comment:
We're using `application/vnd.kubernetes.protobuf;stream=watch` for protobufs.
We should make it consistent.

Reply:
In practice it may not matter much -- I don't know if we expect to serve generic user agents -- but `application/cbor-seq` is registered with the IANA for CBOR Sequences. I think it would be nicer to use this one if we actively adopt CBOR Sequences for this case.